src/dispatcher.rs

```diff
@@ -0,0 +1,110 @@
+use crate::{
+    cache::{self, MessageCache, RoleIdCache},
+    commands::Commands,
+};
+use serenity::{model::prelude::*, prelude::*};
+
+type Result = crate::commands::Result<()>;
+
+pub(crate) struct MessageDispatcher {
+    cmds: Commands,
+}
+
+/// # Dispatcher
+///
+/// This is the event handler for all messages.   
+impl MessageDispatcher {
+    pub fn new(cmds: Commands) -> Self {
+        Self { cmds }
+    }
+}
+
+impl EventHandler for MessageDispatcher {
+    fn message(&self, cx: Context, msg: Message) {
+        self.cmds.execute(cx, msg);
+    }
+
+    fn ready(&self, _: Context, ready: Ready) {
+        println!("{} connected", ready.user.name);
+    }
+}
+
+pub(crate) struct EventDispatcher;
+
+impl RawEventHandler for EventDispatcher {
+    fn raw_event(&self, cx: Context, event: Event) {
+        match event {
+            Event::GuildCreate(ref ev) => {
+                if let Err(e) = init(ev) {
+                    println!("{}", e);
+                }
+            }
+            Event::ReactionAdd(ref ev) => {
+                if let Err(e) = assign_talk_role(&cx, ev) {
+                    println!("{}", e);
+                }
+            }
+            _ => (),
+        }
+    }
+}
+
+fn init(ev: &GuildCreateEvent) -> Result {
```

Comment:
This should operate by Role IDs. Need to grab them from Discord. Could store them as constants, or maybe set them via envvars.

Reply:
Done